File: sstf.c
/*
  SSTF Disk Scheduling Algorithm
*/

#include <stdio.h>
#include <stdlib.h>
#include <math.h>

int main() {
  int queue[100], queue2[100], q_size, head, seek=0, temp;
  float avg;

  printf("%s\n", "-----SSTF Disk Scheduling Algorithm-----");

  printf("%s\n", "Enter the size of the queue");
  scanf("%d", &q_size);

  printf("%s\n", "Enter queue elements");
  for(int i=0; i<q_size; i++){
    scanf("%d",&queue[i]);
  }

  printf("%s\n","Enter initial head position");
  scanf("%d", &head);

  //get distance from head of elems in queue
  for(int i=0; i<q_size; i++){
    queue2[i] = abs(head-queue[i]);
  }

  //swap elems based on their distance from each other
  for(int i=0; i<q_size; i++){
      for(int j=i+1; j<q_size;j++){

        if(queue2[i]>queue2[j]){
            temp = queue2[i];
            queue2[i]=queue[j];
            queue2[j]=temp;

            temp=queue[i];
            queue[i]=queue[j];
            queue[j]=temp;
        }
    }

  }

  for(int i=1; i<q_size; i++){
    seek = seek+abs(head-queue[i]);
    head = queue[i];
  }

  printf("\nTotal seek time is %d\t",seek);
  avg = seek/(float)q_size;
  printf("\nAverage seek time is %f\t", avg);

  return 0;
}

File: disk_sch_java/src/SSTF.java
import java.util.Scanner;

public class SSTF {

    public static void main(String [] args){
        Scanner input  = new Scanner(System.in);
        int [] queue = new int[100];
        int [] queue2 = new int[100];

        int head;
        int seek = 0;
        int qSize;
        int temp;
        float avg;

        System.out.println("SSTF Disk Scheduling");

        System.out.println("Enter queue size");
        qSize = Integer.parseInt(input.nextLine());

        System.out.println("Input Elements");
        for(int i=0; i<qSize;i++){
            queue[i] = Integer.parseInt(input.nextLine());
        }

        System.out.println("Enter initial head position");
        head = Integer.parseInt(input.nextLine());

        for(int i=0; i<qSize; i++){
            queue2[i] = Math.abs(head-queue[i]);
        }

        for(int i=0; i<qSize; i++){

            for(int j=i+1; j<qSize; j++){

                if(queue2[i] >queue2[j]){
                    temp = queue2[i];
                    queue2[i] = queue2[j];
                    queue2[j] = temp;

                    temp = queue[i];
                    queue[i] = queue[j];
                    queue[j] = temp;
                }

            }
        }

        for(int i=1; i<qSize; i++){
            seek = seek + Math.abs(head-queue[i]);
            head = queue[i];
        }

        System.out.println("Total seek time is " +seek);
        avg = seek/(float)qSize;
        System.out.println("Average seek time is " +avg);

    }
}
